File: cpp/cpp_1-9/7.cc
/*
 * C++
 * leetcode7
 * url: https://leetcode-cn.com/problems/reverse-integer/
 * 用long来存答案，如果最后越界了则返回0
 * 如果x是负数，把x弄成正数，flag记为true，最后依据flag将x调回负数
 * 每次薅掉x的最后一位，并加到res 的最后一位
 */

class Solution {
public:
    int reverse(int x) {
    	long res=0;
        long xx = (long)x;
    	bool flag=false;
    	if(xx<0){
    		xx = -xx;
    		flag = true;
    	}
    	while(xx>0){
    		res = res*10+xx%10;
    		xx = xx/10;
    	}
    	if(flag)
    		res = -res;
    	if(res<-2147483648 || res>2147483647)
    		return 0;
    	return (int)res;
    }
};

File: Go/Go_1-9/7.go
// Go
// leetcode 7
// https://leetcode.cn/problems/reverse-integer/

import "math"

func reverse(x int) int {
	var after, before int64
	before = int64(x)
	var isNegative bool
	if before < 0 {
		isNegative = true
		before = -before
	}
	for before > 0 {
		after = after*10 + before%10
		before /= 10
	}
	if isNegative {
		after = -after
	}
	// fmt.Println(after)
	if after >= int64(math.MinInt32) && after <= int64(math.MaxInt32) {
		return int(after)
	}
	return 0
}